src/wp-includes/rest-api/endpoints/class-wp-rest-font-families-controller.php

```diff
@@ -0,0 +1,554 @@
+<?php
+/**
+ * REST API: WP_REST_Font_Families_Controller class
+ *
+ * @package WordPress
+ * @subpackage REST_API
+ * @since 6.5.0
+ */
+
+/**
+ * Font Families Controller class.
+ *
+ * @since 6.5.0
+ */
+class WP_REST_Font_Families_Controller extends WP_REST_Posts_Controller {
+	/**
+	 * Whether the controller supports batching.
+	 *
+	 * @since 6.5.0
+	 * @var false
+	 */
+	protected $allow_batch = false;
+
+	/**
+	 * Checks if a given request has access to font families.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- required by parent class
+		$post_type = get_post_type_object( $this->post_type );
+
+		if ( ! current_user_can( $post_type->cap->read ) ) {
+			return new WP_Error(
+				'rest_cannot_read',
+				__( 'Sorry, you are not allowed to access font families.' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Checks if a given request has access to a font family.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_item_permissions_check( $request ) {
+		$post = $this->get_post( $request['id'] );
+		if ( is_wp_error( $post ) ) {
+			return $post;
+		}
+
+		if ( ! current_user_can( 'read_post', $post->ID ) ) {
+			return new WP_Error(
+				'rest_cannot_read',
+				__( 'Sorry, you are not allowed to access this font family.' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validates settings when creating or updating a font family.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string          $value   Encoded JSON string of font family settings.
+	 * @param WP_REST_Request $request Request object.
+	 * @return true|WP_Error True if the settings are valid, otherwise a WP_Error object.
+	 */
+	public function validate_font_family_settings( $value, $request ) {
+		$settings = json_decode( $value, true );
+
+		// Check settings string is valid JSON.
+		if ( null === $settings ) {
+			return new WP_Error(
+				'rest_invalid_param',
+				/* translators: %s: Parameter name: "font_family_settings". */
+				sprintf( __( '%s parameter must be a valid JSON string.' ), 'font_family_settings' ),
+				array( 'status' => 400 )
+			);
+		}
+
+		$schema   = $this->get_item_schema()['properties']['font_family_settings'];
+		$required = $schema['required'];
+
+		if ( isset( $request['id'] ) ) {
+			// Allow sending individual properties if we are updating an existing font family.
+			unset( $schema['required'] );
```

Comment:
This value is unset here, but it has already been captured into `$required`. Later on, at the end of the method, we iterate over it using `foreach ( $required as $key )`. Is this all intended, or is there a bug hidden here? For example, was the intention actually something like:

```diff
		// Check that none of the required settings are empty values.
-		foreach ( $required as $key ) {
+		if ( isset( $schema['required'] ) ) {
+			foreach ( $required as $key ) {
```

Reply:
To walk through what's going on here, currently:

If a new font family is being created

- `$schema['required']` stays set
- `rest_validate_value_from_schema` validates that those required properties are set (but does not error when they are empty)
- `foreach ( $required as $key ) { ...` checks that required properties are not empty strings

If an existing font family is being updated ( `$request['id']` is set)

- `$schema['required']` is unset, because individual properties (except for slug) can be updated, so no properties are required
- `rest_validate_value_from_schema` validates that properties in the request match the schema (but does not check for required properties)
- `foreach ( $required as $key ) { ...` checks that any schema required properties that are set in the request are not empty

> Is this all intended, or is there a bug hidden here?

It's intended. There are a number of unit tests covering font family creation and updating, which should catch any logic errors here (See `Tests_REST_WpRestFontFamiliesController::test_create_item_invalid_settings` and `Tests_REST_WpRestFontFamiliesController::test_update_item_empty_settings`).

Reply:
Thanks for the comprehensive answer